### DISCUSSION TITLE
Has / Search for custom properties attached to Vertex

### DISCUSSION THREAD

**Discussion by RadekVoltr** [Gremlinq/ExRam.Gremlinq#611] [Q&A]
We are looking for way how to combine strong typed classes with few custom properties.

There is no issue to add them. We are able to read them (not too much pleasant but we can live with it).

But we are unable to find way how to "Where" them. Found in some discussion notice about Properties("prop_name") but Properties with string tag is not allowed by compiler (version 9.2.0). 

Is there way how to send _Has_ command through library which will be based on property not included in class ? 

We are going to have most of of properties in classes but we need to allow user to add optional custom data which are unknown and we need to be able search for them. If there is any other idea which we are missing, I would be happy to hear it.

**Comment by danielcweber** [comment]
You're right, there is no convenient way. I always found that raw string keys would somehow defeat the purpose of an object-graph-mapper. I might reconsider. In any case, it's not a priority for us. You could, however, work around this by adding a custom step/extension method.

**Reply by RadekVoltr** [reply]
Ok, understand.

Can you give me some direction/code where to look/start (or even some example) for adding custom step/extension method ?

one more question - is Gremlin.NET client accessible in case that it is created via .UseCosmosDb(configurator => configurator as part of  IGremlinQuerySource buid ? Don't want to start second one and direct call capability can be handy last resort solution.

**Comment by danielcweber** [comment]
Look at `AsAdmin` yielding an `IGremlinQueryAdmin` for an `AddStep` method. Add custom steps here. For serialization, everything is in a class called `GremlinQueryFragmentSerializer`. Should seem straighforward from there I hope. There's no examples though for this rather advanced feature I am afraid.

As for Gremlin.NET, the client is not eposed anywhere. You'd have to configure it yourself.

**Comment by danielcweber** [comment]
Serialization of steps: https://github.com/ExRam/ExRam.Gremlinq/blob/6b8e980bae1c592e6a28069551323439350a7d34/src/ExRam.Gremlinq.Core/Serialization/GremlinQueryFragmentSerializer.cs#L114

You'd configure the serializer (and override the parts you need) in `ConfigureEnvironment`.`ConfigureSerializer` etc.
